File: jfingerprintmanager/src/main/java/com/jesusm/jfingerprintmanager/authentication/presenter/FingerprintAuthenticationDialogPresenter.java
package com.jesusm.jfingerprintmanager.authentication.presenter;


import android.support.annotation.NonNull;
import android.support.v4.hardware.fingerprint.FingerprintManagerCompat;

import com.jesusm.jfingerprintmanager.base.FingerprintBaseDialogPresenter;

import static com.jesusm.jfingerprintmanager.authentication.presenter.FingerprintAuthenticationDialogPresenter.AuthenticationStage.NEW_FINGERPRINT_ENROLLED;
import static com.jesusm.jfingerprintmanager.authentication.presenter.FingerprintAuthenticationDialogPresenter.AuthenticationStage.PASSWORD;

public class FingerprintAuthenticationDialogPresenter extends FingerprintBaseDialogPresenter {

    /**
     * Enumeration to indicate which authentication method the user is trying to authenticate with.
     */
    enum AuthenticationStage implements FingerprintBaseDialogPresenter.Stage {
        NEW_FINGERPRINT_ENROLLED {
            @Override
            public String id() {
                return "new_fingerprint";
            }
        },
        PASSWORD {
            @Override
            public String id() {
                return "password";
            }
        }
    }

    @NonNull
    protected View view;

    public FingerprintAuthenticationDialogPresenter(@NonNull View view) {
        super(view);
        this.view = view;
    }

    @Override
    public void onAuthenticationSucceeded(FingerprintManagerCompat.AuthenticationResult result) {
        view.onAuthenticationSucceed();
        close();
    }

    @Override
    public void onViewShown() {
        if (!fingerprintHardware.isFingerprintAuthAvailable()) {
            setStage(PASSWORD);
        }

        super.onViewShown();
    }

    @Override
    protected void updateStage() {
        String stageId = stage.id();
        if (stageId.equals(NEW_FINGERPRINT_ENROLLED.id()) || stageId.equals(PASSWORD.id())) {
            goToPassword();
            return;
        }

        super.updateStage();
    }

    private void goToPassword() {
        showPassword();
        cancelFingerprintAuthenticationListener();
    }

    private void showPassword() {
        view.onPasswordViewDisplayed(stage == NEW_FINGERPRINT_ENROLLED);
    }

    public void showPasswordClicked() {
        setStage(PASSWORD);
        updateStage();
    }

    public void onPasswordEntered(String password, boolean useFingerprintFuture) {
        if (stage == BaseStage.FINGERPRINT) {
            goToPassword();
        } else {
            verifyPassword(password, useFingerprintFuture);
        }
    }

    private void verifyPassword(String password, boolean useFingerprintFuture) {
        if (!isValidPassword(password)) {
            view.onPasswordEmpty();
            return;
        }

        if (stage == NEW_FINGERPRINT_ENROLLED) {
            view.saveUseFingerprintFuture(useFingerprintFuture);
            if (useFingerprintFuture) {
                // Re-create the key so that fingerprints including new ones are validated.
                view.createKey();
                setStage(BaseStage.FINGERPRINT);
            }
        }

        view.onPasswordInserted(password);
        close();
    }

    private boolean isValidPassword(String password) {
        return password != null && password.length() > 0;
    }

    public void newFingerprintEnrolled() {
        stage = NEW_FINGERPRINT_ENROLLED;
    }

    public interface View extends FingerprintBaseDialogPresenter.View {
        void saveUseFingerprintFuture(boolean useFingerprintFuture);

        void createKey();

        void onPasswordInserted(String password);

        void onPasswordEmpty();

        void onAuthenticationSucceed();

        void onPasswordViewDisplayed(boolean newFingerprintEnrolled);
    }
}

File: kfingerprintmanager/src/main/kotlin/com/jesusm/kfingerprintmanager/authentication/presenter/FingerprintAuthenticationDialogPresenter.kt
package com.jesusm.kfingerprintmanager.authentication.presenter

import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
import com.jesusm.kfingerprintmanager.base.ui.presenter.FingerprintBaseDialogPresenter

class FingerprintAuthenticationDialogPresenter(view: View) : FingerprintBaseDialogPresenter(view) {
    fun onPasswordEntered(password: String, useFingerprintFuture: Boolean) {
        if (stage === FingerprintBaseDialogPresenter.Stage.FINGERPRINT) {
            goToPassword()
        } else {
            verifyPassword(password, useFingerprintFuture)
        }
    }

    private fun goToPassword() {
        showPassword()
        cancelFingerprintAuthenticationListener()
    }

    private fun verifyPassword(password: String, useFingerprintFuture: Boolean) {
        if (!isValidPassword(password)) {
            (view as View).onPasswordEmpty()
            return
        }

        if (stage === Stage.NEW_FINGERPRINT_ENROLLED) {
            (view as View).saveUseFingerprintFuture(useFingerprintFuture)
            if (useFingerprintFuture) {
                // Re-create the key so that fingerprints including new ones are validated.
                view.createKey()
                stage = FingerprintBaseDialogPresenter.Stage.FINGERPRINT
            }
        }

        close()
        (view as View).onPasswordInserted(password)
    }

    private fun isValidPassword(password: String) : Boolean = password.isNullOrEmpty().not()

    private fun showPassword() {
        (view as View).onPasswordViewDisplayed(stage == Stage.NEW_FINGERPRINT_ENROLLED)
    }

    fun showPasswordClicked() {
        stage = Stage.PASSWORD
        updateStage()
    }

    fun newFingerprintEnrolled() {
        stage = Stage.NEW_FINGERPRINT_ENROLLED
    }

    override fun onViewShown() {
        if (!fingerprintHardware.isFingerprintAuthAvailable()) {
            stage = Stage.PASSWORD
        }

        super.onViewShown()
    }

    override fun updateStage() {
        when(stage)
        {
            Stage.NEW_FINGERPRINT_ENROLLED -> goToPassword()
            Stage.PASSWORD -> goToPassword()
            else -> super.updateStage()
        }
    }

    override fun onAuthenticationSucceeded(result: FingerprintManagerCompat.AuthenticationResult) {
        close()
        (view as View).onAuthenticationSucceed()
    }

    interface View : FingerprintBaseDialogPresenter.View {
        fun saveUseFingerprintFuture(useFingerprintFuture: Boolean)

        fun createKey()

        fun onPasswordInserted(password: String)

        fun onPasswordEmpty()

        fun onAuthenticationSucceed()

        fun onPasswordViewDisplayed(newFingerprintEnrolled: Boolean)
    }
}